src/uct/ib/base/ib_iface.c

```diff
@@ -667,6 +667,60 @@ uct_ib_iface_roce_is_reachable(const uct_ib_device_gid_info_t *local_gid_info,
     return ret;
 }
 
+static int
+uct_ib_iface_is_same_device(const uct_ib_address_pack_params_t *params,
+                            uct_ib_iface_t *iface)
+{
+    const union ibv_gid *gid = &iface->gid_info.gid;
+    int res = 1;
+
+    if ((params->flags & UCT_IB_ADDRESS_PACK_FLAG_ETH) !=
+        uct_ib_iface_is_roce(iface)) {
+        return 0;
+    }
+
+    if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_ETH) {
+        res &= !memcmp(gid->raw, params->gid.raw, sizeof(params->gid.raw));
+    } else {
+        res &= (uct_ib_iface_port_attr(iface)->lid == params->lid);
+        if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID &&
+            params->gid.global.interface_id != 0) {
+            res &= (params->gid.global.interface_id ==
+                    gid->global.interface_id);
+        }
+        if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_SUBNET_PREFIX) {
+            res &= (params->gid.global.subnet_prefix ==
+                    gid->global.subnet_prefix);
+        }
+    }
+
+    if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_PKEY) {
+        res &= (params->pkey == iface->pkey);
+    }
+
+    return res;
+}
+
+int uct_ib_iface_is_reachable_v2(const uct_iface_h tl_iface,
+                                 const uct_iface_is_reachable_params_t *params)
+{
+    uct_ib_iface_t *iface           = ucs_derived_of(tl_iface, uct_ib_iface_t);
+    const uct_ib_address_t *ib_addr = (const void*)params->device_addr;
+    uct_ib_address_pack_params_t pack_params;
+    uct_iface_reachability_scope_t scope;
+
+    scope = UCS_PARAM_VALUE(UCT_IFACE_IS_REACHABLE_FIELD, params, scope, SCOPE,
+                            UCT_IFACE_REACHABILITY_SCOPE_NETWORK);
+
+    if (scope == UCT_IFACE_REACHABILITY_SCOPE_DEVICE) {
+        uct_ib_address_unpack(ib_addr, &pack_params);
+        return uct_ib_iface_is_same_device(&pack_params, iface);
+    }
+
+    return uct_ib_iface_is_reachable(tl_iface, params->device_addr,
+                                     params->iface_addr);
```

Comment:
we can call uct_ib_iface_is_reachable anyway. if it fails - it's definitely not same device.
then we can skip some checks in uct_ib_iface_is_same_device